[SPI Checking] Add build-time tool for auditing SPI use

emw-apple opened the pull request:
#### 0a497503904f1efffa0a777a416bfa2224f0d6f5
<pre>
[SPI Checking] Add build-time tool for auditing SPI use
<a href="https://bugs.webkit.org/show_bug.cgi?id=290424">https://bugs.webkit.org/show_bug.cgi?id=290424</a>
<a href="https://rdar.apple.com/135979752">rdar://135979752</a>

Reviewed by Sam Sneddon and BJ Burg.

Introduce the `webkitapipy` library and the audit-spi tool for detecting
SPI use in Mach-O binaries via static analysis. Integrate audit-spi into
the build system but leave it off by default, for now.

BACKGROUND
==========

Binary analysis tools (dyld_info) provide a record of the symbols and
Objective-C metadata used by a Mach-O binary. For each of those records,
webkitapipy checks whether the declaration is known to be API.

To do this, it uses partial SDKDBs built from projects in the SDK.
SDKDBs are a machine-readable representation of all the symbolic and
language-level declarations in a project, which are generated by TAPI.

Since the SDK is relatively unchanging compared to local build products,
we do not attempt to produce SDKDB records for the whole SDK at build
time. In a follow-up commit, WebKit will check in partial SDKDBs
generated from supported public SDKs using metadata from Apple&apos;s
internal build system. For checking internal SDKs, Apple engineers will
use separate SDKDB records available internally.

PROCEDURE
=========

At a high level, webkitapipy has a main entrypoint (program.py) which
takes a directory of SDKDBs and binaries in a build directory, and
performs SPI analysis in three stages:

1. Build an internal cache of all symbols, classes, and selectors in
   the SDK. This representation supports fast querying and is reusable
   in incremental builds. It keeps track of input files that were used to
   seed the cache along, and updates cache information when input files
   change. It is backed by a sqlite database stored in the build
   directory.

2. Load symbol, classes, and selectors from locally-built frameworks
   and dylibs whose SPI (and IPI) are allowed to be used from the binary
   being analyzed. These are also stored in the cache.

3. Run nm and various otool invocations on the binary being analyzed to
   get its symbol table, referenced selectors, and soft-linking
   metadata. Check all of these records against the cache. Missing
   records are considered to be potential SPI and are reported.

Future work will add a fourth stage, which is to parse and consult an
SPI allowlist. The eventual goal is to make use of un-allowlisted SPI a
build error, requiring committers to keep track of any SPI usage they
introduce.

INTEGRATION
===========

In the build, WebKit&apos;s frameworks have an &quot;Audit SPI use&quot; build phase
that calls into audit-spi. It is controlled by a WK_ANALYZE_SPI build
setting. Currently it off by default, but will eventually be
enabled for debug iOS configurations.

IMPLEMENTATION NOTES
====================

As a new, relativly self-contained Python module, I&apos;ve taken the
opportunity to adopt type hints in webkitapipy. The code typechecks
successfully with mypy (`python3 -m mypy -p webkitapipy
--check-untyped-defs`). It is also fully PEP8 compliant.

A follow-up change will add a unit test which runs mypy on the module. I
hope other contributors find it helpful to adopt type hinting in their
libraries and make it easier to catch bugs ahead of time.

* Configurations/CommonBase.xcconfig:
* Configurations/SPIAnalysis.xcconfig: Added.
* Source/JavaScriptCore/JavaScriptCore.xcodeproj/project.pbxproj:
* Source/WTF/Scripts/audit-spi-if-needed.sh: Added. The Xcode frontend
  to audit-spi, which checks build settings and invokes the tool from
  the build environment. It is present in production builds (i.e.
  outside of the Tools/ directory) so that the script phases have
  something to call into.
* Source/WTF/WTF.xcodeproj/project.pbxproj:
* Source/WebCore/Configurations/WebCore.xcconfig:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebGPU/WebGPU.xcodeproj/project.pbxproj:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKitLegacy/WebKitLegacy.xcodeproj/project.pbxproj:
* Tools/Scripts/audit-spi: Added.
* Tools/Scripts/webkitpy/__init__.py: Move the PyYAML autoinstall
  dependency here, since it is no longer just used for buildbot.
* Tools/Scripts/libraries/webkitapipy/__init__.py: Added.
* Tools/Scripts/libraries/webkitapipy/program.py: Added. Implementation
  of the audit-spi tool. Handles argument parsing, loading the SDKDB
  and binary report, and reporting issues.
(TSVReporter): Very simple base class for reporting missing API as
tabular data. Will be subclasses and/or replaced in the future with
implementation that manages the SPI allowlist.
(TSVReporter.__init__):
(TSVReporter.missing_selector):
(TSVReporter.missing_class):
(TSVReporter.missing_symbol):
(TSVReporter.finished):
(main):
* Tools/Scripts/libraries/webkitapipy/macho.py: Added. Interface for
  extracting symbol and ObjC metadata from binaries and producing an
  APIReport.
(APIReport):
(APIReport.create):
(APIReport._populate_from_dyld_info):
* Tools/Scripts/libraries/webkitapipy/sdkdb.py: Added. The
  implementation of the sqlite-backed SDKDB cache.
(sdkdb):
(sdkdb.__init__):
(sdkdb._initialize_db):
(sdkdb.__del__):
(sdkdb.__enter__):
(sdkdb.__exit__):
(sdkdb._cache_hit_preparing_to_insert):
(sdkdb.add_partial_sdkdb):
(sdkdb.add_symbols):
(sdkdb.add_tbd):
(sdkdb.symbol):
(sdkdb.objc_class):
(sdkdb.objc_class.WHERE):
(sdkdb.objc_selector):
(sdkdb.stats):
(sdkdb._add_objc_interface):
(sdkdb._add_symbol):
(sdkdb._add_objc_class):
(sdkdb._add_objc_class.VALUES):
(sdkdb._add_objc_selector):
* Tools/Scripts/libraries/webkitapipy/tbd.py: Added. In addition to
  loading partial SDKDBs generated from the SDK, there are some
  libraries in the OS which are implicitly treated as API, such as
  libSystem. For these libraries, webkitapipy imports symbol
  information directly from TBDs in the SDK.
(ExportList):
(ExportList.from_json):
(TBD):
(TBD.from_json):
(TBD.from_file): Use llvm-readtapi to convert a SDKDB (typically
YAML-formatted) into a JSON representation.
* Tools/Scripts/webkitpy/autoinstalled/buildbot.py: Move PyYaml
  autoinstaller declaration from here.

Canonical link: <a href="https://commits.webkit.org/293695@main">https://commits.webkit.org/293695@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3dec077adb4c719779eb9bcca4395a3823beddd4

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/99645 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/19295 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/9556 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/104775 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/50241 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/101686 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/19584 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/27728 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/75852 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/32948 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/102652 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/14917 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/89979 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/56211 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/99118 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/14717 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/7965 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/49603 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/92325 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/84679 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/8050 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/107134 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/98263 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/26760 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/19537 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/84806 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/27124 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/86183 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/84325 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/21406 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/29006 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/6712 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/20550 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/26700 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/31902 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/121879 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/26515 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/34040 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/29831 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/28085 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

gsnedders commented on Tools/Scripts/webkitpy/api_analyzer/driver.py:
```diff
@@ -0,0 +1,241 @@
+from importlib.machinery import FileFinder
+from pathlib import Path
+
+from webkitpy import api_analyzer
```
This is literally importing itself?

emw-apple replied:
I need to get the package object (for dependency tracking below).

gsnedders commented on Tools/Scripts/webkitpy/api_analyzer/driver.py:
```diff
@@ -0,0 +1,241 @@
+        self.seen: set[str] = set()
+        self.print_details = args.details
+
+    def missing_selector(self, name: str, ignore=False):
```
Let's make these boolean arguments all keyword-only:

```suggestion
    def missing_selector(self, name: str, *, ignore=False):
```

gsnedders commented on Tools/Scripts/webkitpy/api_analyzer/driver.py:
```diff
@@ -0,0 +1,241 @@
+    parser.add_argument('--depfile', type=Path,
+                        help='write inputs used for incremental rebuilds')
+    parser.add_argument('--details', action='store_true',
+                        help='print a line for each unknown symbol')
```
I'd split this into a separate `get_parser` function.

gsnedders commented on Tools/Scripts/webkitpy/api_analyzer/driver.py:
```diff
@@ -0,0 +1,241 @@
+    parser.add_argument('--details', action='store_true',
+                        help='print a line for each unknown symbol')
+    if additions:
+        additions.add_arguments(parser)
```
We might want to pass an already constructed parser as a parent to the main parser, c.f. https://docs.python.org/3/library/argparse.html#parents, instead of adding them like this?

gsnedders commented on Tools/Scripts/webkitpy/api_analyzer/driver.py:
```diff
@@ -0,0 +1,241 @@
+                        help='binaries and SDKDBs to allow arbitrary use of')
+    parser.add_argument('-a', '--arch-name', required=True,
+                        help='which architecture to analyze binary with')
+    parser.add_argument('--primary-file', help='file to analyze', type=Path)
```
"primary" implies there might be others, which the help string doesn't?

I presume this is probably mostly about C pre-processor includes?

emw-apple replied:
The others are the `input_files`. This is a swift compiler nomenclature, not that that makes it any clearer :)

What I'm trying to get at is that the `primary_file` is the binary we're analyzing for SPI use. The other `input_files` are binaries that the primary file is allowed arbitrary use of. For example, when the primary file is WebKit.framework/WebKit, input files are WebCore, JavaScriptCore, etc.

Do you have a preference on how to express that?

gsnedders commented on Tools/Scripts/webkitpy/api_analyzer/driver.py:
```diff
@@ -0,0 +1,241 @@
+    args = parser.parse_args(argv)
+
+    # For the depfile, start with the paths of all the modules in api_analyzer
+    # since this library is part of WebKit source code.
```
This feels kinda surprising, but I guess is reasonable… It kinda just feels like surprising meta-programming, but probably is reasonable?

emw-apple replied:
I'm open to being told that it's too magical, but we do need to produce a list of the names of these files _somewhere_. It could be a handwritten list, or part of the `analyze-spi.sh` script that the build system invokes, or names could be listed directly in Xcode.

gsnedders commented on Tools/Scripts/webkitpy/api_analyzer/driver.py:
```diff
@@ -0,0 +1,241 @@
+    # For the depfile, start with the paths of all the modules in api_analyzer
+    # since this library is part of WebKit source code.
+    inputs = []
+    for info in pkgutil.walk_packages(api_analyzer.__path__):
```
Don't need to explicitly specify the module, because the module scope is our global scope. Also, probably better to pass the name too (which only really matters if we're a module and not a package, but it doesn't hurt):

```suggestion
    for info in pkgutil.walk_packages(__path__, __name__ + "."):
```

emw-apple replied:
I played around with this but I'm not following. To start, `__path__` isn't defined because this code isn't at the top-level of a package.

Latest revision is a little more explicit about what's happening, and is also tracking source code dependencies for `webkitapipy_additions`.

gsnedders commented on Tools/Scripts/webkitpy/api_analyzer/driver.py:
```diff
@@ -0,0 +1,241 @@
+    inputs = []
+    for info in pkgutil.walk_packages(api_analyzer.__path__):
+        # This code asssumes api_analyzer is a file-based module.
+        assert isinstance(info.module_finder, FileFinder)
```
Do we actually care? We only care that origin is not `None`.

emw-apple replied:
I don't care, but mypy does…I'll see if I can find a better way to express it.

gsnedders commented on Tools/Scripts/webkitpy/api_analyzer/driver.py:
```diff
@@ -0,0 +1,241 @@
+        if n_changes % 10 == 0:
+            print(f'{n_changes} projects...')
+
+    db_initialization_start = time.time()
```
We probably want `time.monotonic` or maybe even `time.perf_counter` instead.

gsnedders commented on Tools/Scripts/webkitpy/api_analyzer/driver.py:
```diff
@@ -0,0 +1,241 @@
+}
+
+
+class TSVReporter:
```
I wonder if we should use the `csv` module for output from this, hmm? Probably not worthwhile given complexity here.

emw-apple replied:
Yeah, I started down that path and decided it wasn't actually worth it. Also, I'm not sure whether the tabular output format will stick around long-term, since the goal is for this to eventually be an allowlist checker. (But even then, being able to spit out tabular data is nice.)

gsnedders commented on Tools/Scripts/webkitpy/api_analyzer/driver.py:
```diff
@@ -0,0 +1,241 @@
+
+
+class TSVReporter:
+    def __init__(self, args: argparse.Namespace):
```
I'm never the biggest fan of passing around `argparse.Namespace` everywhere; it's untyped and makes testing harder.

emw-apple replied:
What I want here is a quick-and-dirty way for the `webkitapipy_additions` library to be able to add internal-only arguments that influence its reporter.

gsnedders commented on Tools/Scripts/webkitpy/api_analyzer/driver.py:
```diff
@@ -0,0 +1,241 @@
+
+
+if __name__ == '__main__':
+    main()
```
We might want this file to actually be called `__main__.py`, c.f. https://docs.python.org/3/library/__main__.html#main-py-in-python-packages

emw-apple replied:
TIL. Yeah, that could be nice, and it's also kinda overkill since I ended up writing an executable stub in `Tools/Scripts/analyze-spi`.

emw-apple replied:
I ended up removing this entirely, since we now have a formal entrypoint in `pyproject.toml`.

gsnedders commented on Tools/Scripts/webkitpy/api_analyzer/driver.py:
```diff
@@ -0,0 +1,241 @@
+        with open(args.depfile, 'w') as fd:
+            fd.write('dependencies: ')
+            fd.write(' \\\n  '.join(shlex.quote(os.path.abspath(path))
+                                    for path in inputs))
```
No trailing new line, given UNIX definitions of what a text file is?

gsnedders commented on Tools/Scripts/webkitpy/api_analyzer/driver.py:
```diff
@@ -0,0 +1,241 @@
+            ignored = any(fnmatch(symbol, pattern)
+                          for pattern in ALLOWED_SYMBOL_GLOBS)
+        if symbol.startswith('_OBJC_CLASS_$_'):
+            class_name = symbol[14:]
```
`symbol.removeprefix('_OBJC_CLASS_$_')` would be clearer

gsnedders commented on Tools/Scripts/webkitpy/api_analyzer/driver.py:
```diff
@@ -0,0 +1,241 @@
+        if not db.objc_selector(selref) and selref not in report.selectors:
+            reporter.missing_selector(selref)
+
+    for symbol in sorted(report.bindings):
```
These two loops look like they maybe should be in a `def process_report(reporter: Reporter, db: SDKDB) -> None` so they can be tested in isolation to the rest?

gsnedders commented on Tools/Scripts/webkitpy/api_analyzer/macho.py:
```diff
@@ -0,0 +1,128 @@
+
+def report(binary_path: Path, arch: str, symbols_only=False) -> Report:
+    # Symbol table
+    nm_lines = subprocess.check_output(
```
We should be using `subprocess.run` nowadays in preference to this, with `check=True` and `capture_output=True`.

emw-apple replied:
At risk of nitpicking, why? Seems like it's more to type and expresses the exact same behavior.

gsnedders replied:
They're specifically documented as the "older high-level API" in the docs, which makes me suspect it'll get deprecated and removed.

gsnedders commented on Tools/Scripts/webkitpy/api_analyzer/macho.py:
```diff
@@ -0,0 +1,128 @@
+    # Symbol table
+    nm_lines = subprocess.check_output(
+        ('xcrun', 'nm', '-P', '--arch', arch, binary_path),
+        text=True, errors='replace'
```
What output are we getting that fails to decode as UTF-8? What encoding is it in?

emw-apple replied:
Unknown, but IIUC Mach-O does not require symbol names to be UTF-8, so I worry about encoding errors coming from arbitrary non-API-related symbols in a binary.

emw-apple replied:
It doesn't hurt anything to remove this, so I'll take it out.

gsnedders commented on Tools/Scripts/webkitpy/api_analyzer/sdkdb.py:
```diff
@@ -0,0 +1,261 @@
+VERSION = 1
+
+
+class sdkdb:
```
PascalCase would be the norm for a class, or maybe even uppercase given it's a wrapper around an SDKDB?

gsnedders reviewed (commented):
This is only a partial look, given I have a medical appointment to go to. Some of these comments could easily have a "ditto" added to other similar code.

My only really high-level question is "should this be in Tools/Scripts/libraries and not Tools/Scripts/webkitpy", given it doesn't really depend on `webkitpy` in any way (and it would be nice to keep it that way).

burg commented on Configurations/SPIAnalysis.xcconfig:
```diff
@@ -0,0 +1,39 @@
+
+// SPI analysis is only enabled when building for iOS in debug-style builds.
+// FIXME: Uncomment to enable by default.
+//WK_ANALYZE_SPI = $(WK_ANALYZE_SPI_$(WK_PLATFORM_NAME));
```
Can we factor this so that we don't leave commented lines in, and instead folks can use LocalOverrides.xcconfig to opt into more processing?

emw-apple replied:
Sure. The intention here is to show what the default will be once I enable checking in a follow-up. But you're right that it's cleaner to avoid proposing currently-off-but-eventually-on-by-default settings here.

burg commented on Source/WebGPU/WebGPU.xcodeproj/project.pbxproj:
```diff
@@ -1205,6 +1206,27 @@
+			);
+			runOnlyForDeploymentPostprocessing = 0;
+			shellPath = /bin/sh;
+			shellScript = "\"${SCRIPT_INPUT_FILE_0}\" \"${BUILT_PRODUCTS_DIR}/JavaScriptCore.framework/JavaScriptCore\"{,.partial.sdkdb} \"${BUILT_PRODUCTS_DIR}/WebGPU.framework/WebGPU.partial.sdkdb\"\n";
```
Is it generally safe to use shell expansion here like this?

emw-apple replied:
I think so? It does lock us into `bash` semantics, but that's what the OS provides.

One enhancement I have in mind is to move some of this logic inside the library, and have the command-line interface take linker-style arguments, so the build system would pass `-framework JavaScriptCore` and it'd know to look for the respective binary and SDKDB. I think I'd prefer to do that in a follow-up change.

burg commented on Source/WTF/Scripts/analyze-spi.sh:
```diff
@@ -0,0 +1,20 @@
+set -e
+
+program="${SRCROOT}/../../../${WK_ADDITIONAL_SCRIPTS_DIR}/analyze-spi"
+[ -f "${program}" ] || program="${SRCROOT}/../../Tools/Scripts/analyze-spi"
```
Is SRCROOT here the SRCROOT for WTF, or is it `Source/WTF/Scripts/` ?

emw-apple replied:
SRCROOT of whichever Xcode project is using the script. For WebKit, all our (non-third-party) projects are two levels down from the workspace root.

(fwiw, I based this on the various `check-xcfilelists.sh` scripts we have in each project)

burg commented on Source/WTF/Scripts/analyze-spi.sh:
```diff
@@ -0,0 +1,20 @@
+    (set -x && "${program}" \
+     --sdkdb-dir "${WK_SDKDB_DIR}/${WK_SDKDB_NAME}" \
+     --sdkdb-cache "${OBJROOT}/WebKitSDKDBs/${WK_SDKDB_NAME}.sqlite3" \
+     --sdk-dir "${SDKROOT}" --arch-name "${ARCHS%% *}" \
```
Can you explain what the `${ARCHS%% *}` does? I can't find info other than that it's a greedy substitution modifier.

emw-apple replied:
It's a bash-ism to get the first whitespace-separate word from a variable. So what's happening here is that if a build is multi-arch, we take the first architecture from the list and use that for analysis.

…which is perhaps a dangerous corner to cut. For example, our macOS CI happened to pass `ARCHS="x86_64 arm64"` we'd only ever check the legacy architecture.

emw-apple replied:
Latest revision checks all architectures, which will be a tiny bit more work to do for CI builders, but will have no impact at desk.

burg commented on Source/WTF/Scripts/analyze-spi.sh:
```diff
@@ -0,0 +1,20 @@
+     --sdkdb-dir "${WK_SDKDB_DIR}/${WK_SDKDB_NAME}" \
+     --sdkdb-cache "${OBJROOT}/WebKitSDKDBs/${WK_SDKDB_NAME}.sqlite3" \
+     --sdk-dir "${SDKROOT}" --arch-name "${ARCHS%% *}" \
+     --depfile "${DERIVED_FILES_DIR}/api-analyzer.d" \
```
should this be `analyze-spi.d` ? Or named after the phase?

burg commented on Source/WebKit/WebKit.xcodeproj/project.pbxproj:
```diff
@@ -19666,6 +19669,69 @@
+			);
+			runOnlyForDeploymentPostprocessing = 0;
+			shellPath = /bin/sh;
+			shellScript = "\"${SCRIPT_INPUT_FILE_0}\" \"${BUILT_PRODUCTS_DIR}/JavaScriptCore.framework/JavaScriptCore\"{,.partial.sdkdb} \"${BUILT_PRODUCTS_DIR}/libwebrtc.dylib\" \"${BUILT_PRODUCTS_DIR}/WebCore.framework/WebCore\"{,.partial.sdkdb} \"${BUILT_PRODUCTS_DIR}/WebGPU.framework/WebGPU\"{,.partial.sdkdb} \"${BUILT_PRODUCTS_DIR}/libWebKitSwift\"{.dylib,.partial.sdkdb} \"${BUILT_PRODUCTS_DIR}/WebKi ...
```
I'm concerned about the maintainability of inlining these shell scripts which seems to mostly be used to expand file paths with shell expansion. Can this be a separate shell script?

emw-apple replied:
They can, although I'd like to avoid needing to check in 5 shell scripts plus the script in WTF that they all call into. If I implement the `-framework ...` style lookup that I described in https://github.com/WebKit/WebKit/pull/43045/files#r2027566275, would that keep things sufficiently terse that you'd be okay having it written inline?

burg replied:
Yeah, I prefer the per-"framework" approach is taken for Inspector/WebDriver code generation. See `generate-inspector-protocol-bindings.py` and its invocations in script phases.

burg commented on Tools/Scripts/libraries/webkitapipy/webkitapipy/macho.py:
```diff
@@ -0,0 +1,128 @@
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+import subprocess
+import re
```
Nit: ordering

burg commented on Tools/Scripts/libraries/webkitapipy/webkitapipy/macho.py:
```diff
@@ -0,0 +1,128 @@
+    file: Path
+
+
+def report(binary_path: Path, arch: str, symbols_only=False) -> Report:
```
Naming nit:  `create_report` or `create_api_report`

burg commented on Tools/Scripts/libraries/webkitapipy/webkitapipy/macho.py:
```diff
@@ -0,0 +1,128 @@
+method_to_selector = re.compile(r'[-+]\[(?P<class>\S+) (?P<selector>[^\]]+)\]')
+
+
+class Report(NamedTuple):
```
Naming nit: would love an `API` or `SPI` or `Usage` in there so that it's clearer what kind of report it represents.

burg commented on Tools/Scripts/libraries/webkitapipy/webkitapipy/macho.py:
```diff
@@ -0,0 +1,128 @@
+        elif flag == 'U':
+            bindings.append(symbol)
+
+    if not symbols_only:
```
Recommend some sort of early return or helper function call - that's a lot of indent.

burg commented on Tools/Scripts/libraries/webkitapipy/webkitapipy/macho.py:
```diff
@@ -0,0 +1,128 @@
+
+    for line in nm_lines:
+        if not line or f'for architecture {arch}' in line:
+            # nm always
```
Not sure what the comment means.

burg commented on Tools/Scripts/libraries/webkitapipy/webkitapipy/macho.py:
```diff
@@ -0,0 +1,128 @@
+        if not line or f'for architecture {arch}' in line:
+            # nm always
+            continue
+        symbol, flag, *_ = line.rsplit(maxsplit=4)
```
TIL about *_ destructuring O_O

burg commented on Tools/Scripts/libraries/webkitapipy/webkitapipy/macho.py:
```diff
@@ -0,0 +1,128 @@
+        for line in it:
+            if line == 'Contents of (__DATA_CONST,__objc_selrefs) section' or \
+                    line == 'Contents of (__DATA,__objc_selrefs) section':
+                for line in it:
```
Note: it gave me a little pause to see nested iteration of the iterator, but I guess there's nothing wrong with it, just unusual.

burg commented on Tools/Scripts/libraries/webkitapipy/webkitapipy/macho.py:
```diff
@@ -0,0 +1,128 @@
+            text=True, errors='replace'
+        ).splitlines()
+
+        it = iter(otool_lines)
```
Really don't like continued reassignments to `it` and `otool_lines`. I think `iter(otool_lines)` could be inlined in most places `it` is being passed rather than used directly.

burg commented on Tools/Scripts/libraries/webkitapipy/webkitapipy/macho.py:
```diff
@@ -0,0 +1,128 @@
+
+def _otool_parse_string_section(it: Iterator[str],
+                                section: str) -> Iterator[str]:
+    # otool's -vV prints raw sections as an address, followed by hex bytes,
```
Thanks for this comment!

burg commented on Tools/Scripts/libraries/webkitapipy/webkitapipy/program.py:
```diff
@@ -0,0 +1,244 @@
+    # Initializing the SDKDB cache from scratch takes some time (~ 15-20 sec).
+    # Print progress updates and measure execution time to indicate how much
+    # build time the cache will save.
+    n_changes = 0
```
Nit: change_count?

burg commented on Tools/Scripts/libraries/webkitapipy/webkitapipy/sdkdb.py:
```diff
@@ -0,0 +1,261 @@
+    and it is re-added.
+
+    To manage concurrent access to a cache, enter the sdkdb as a context
+    manager before using the add_* methods. Prefer grouping
```
Nit: comment seems clipped

emw-apple replied:
Good catch, thank you!

burg reviewed (commented):
This is a really impressive tool. I left a bunch of comments, didn't get too deep into the sqlite code, but otherwise looks good.

aproskuryakov commented on Tools/Scripts/libraries/webkitapipy/webkitapipy/sdkdb_from_build_products.py:
```diff
@@ -0,0 +1,83 @@
+from pathlib import Path
+
+
+def is_executable_file(path):
```
This is a somewhat weak implementation of it, in multiple ways. And I generally wouldn't even think of `.a` or `.dylib` as "executable".

I don't fully understand what you are trying to do, but:
- if possible, make the name more accurate;
- if it's still not 100% descriptive, add some comment that would effectively discourage people from reusing this function in new places.

emw-apple replied:
New name is `belongs_in_sdk_content`, which hopefully makes the purpose of this much clearer.

aproskuryakov replied:
It's better, although still seems misleading. Why don't `.a` files belong in SDK content? We used to have those in some SDK flavors.

aproskuryakov commented on Tools/Scripts/libraries/webkitapipy/webkitapipy/sdkdb_from_build_products.py:
```diff
@@ -0,0 +1,83 @@
+                        help='frameworks, dylibs, or root directories to '
+                        'include in generated sdkdb')
+    parser.add_argument('--sdk-dir',
+                        help='SDK to use when parsing public headers')
```
Might be good to say that it defaults to SDKROOT if available. But what if neither is available? From a cursory glance, looks like the script will misbehave.

aproskuryakov commented on Tools/Scripts/libraries/webkitapipy/webkitapipy/sdkdb_from_build_products.py:
```diff
@@ -0,0 +1,83 @@
+                           '--action=scan-interface',
+                           '--sdk-content-root', sdk_content,
+                           '--runtime-root', runtime_content,
+                           '--sdk', sdk_dir,
```
I don't see an `--sdk` option in help, is this short for `--sdk-content-root`?

emw-apple replied:
Looks like it is an undocumented alias for `-isysroot`: https://github.com/apple-oss-distributions/tapi/blob/b824911e1ab708202a6bc1291b9e9941946db371/include/tapi/Driver/TAPIOptions.td#L457-L458

Don't know how I ended up with it; I'll replace with the documented version.

aproskuryakov commented on Tools/Scripts/libraries/webkitapipy/webkitapipy/sdkdb_from_build_products.py:
```diff
@@ -0,0 +1,83 @@
+        for idx in reversed(range(len(parts))):
+            part = parts[idx]
+            if part.endswith('.framework'):
+                path_in_root = Path('System/Library/PrivateFrameworks',
```
The hardcoded paths feel dangerous, but I don't have a better suggestion. Perhaps this needs some comments, with the expectation that other people will have to update this code in the future?

aproskuryakov commented on Tools/Scripts/libraries/webkitapipy/webkitapipy/sdkdb_from_build_products.py:
```diff
@@ -0,0 +1,83 @@
+            shutil.copytree(product, sdk_content / path_in_root,
+                            ignore=lambda path, contents: [
+                                name for name in contents
+                                if is_executable_file(os.path.join(path, name))
```
Is there a reason why you are mixing `pathlib` and `os.path` (operator `/` above, and `path.join` here)?

emw-apple replied:
Mostly because `shutil.copytree` doesn't pass Path objects into this callable. But I think I prefer it your way, as I think it makes the function above more readable.

aproskuryakov commented on Tools/Scripts/libraries/webkitapipy/webkitapipy/sdkdb_from_build_products.py:
```diff
@@ -0,0 +1,83 @@
+                           '--output', tapi_output.parent))
+    shutil.copyfile(tapi_output, output_file)
+    if not args.output:
+        print('Generated', output_file)
```
Seems like this should have a space between, or better yet, to be an f-string.

emw-apple replied:
`print` is space-separated by default, but an f-string is fine.

aproskuryakov commented on Tools/Scripts/libraries/webkitapipy/webkitapipy/sdkdb_from_build_products.py:
```diff
@@ -0,0 +1,83 @@
+                                if is_executable_file(os.path.join(path, name))
+                            ])
+            shutil.copytree(product, runtime_content / path_in_root)
+        except NotADirectoryError:
```
I'm not a fan of using exceptions for non-exceptional code paths, but I understand that Python directory copying is a mess and this could be the most elegant solution.

emw-apple replied:
I'll replace it with `Path.is_dir()`, which adds a syscall but sounds like it's well worth it for readability.

aproskuryakov reviewed (commented):
I looked at sdkdb_from_build_products.py, and while there are multiple nitpicks, it looks fine to me. Hopefully others can review the rest.

burg reviewed (approved):
I am supportive of moving forward on this new script so we can get more testing underway. But, please refresh the changelog to match the final plan if we are not actually using `nm` and `otool` anymore.

burg commented on Configurations/CommonBase.xcconfig:
```diff
@@ -66,6 +66,10 @@ OTHER_LDFLAGS = $(inherited) $(WK_COMMON_OTHER_LDFLAGS);
 WK_COMMON_OTHER_TAPI_FLAGS = -x objective-c++ -std=c++2b -fno-rtti $(WK_SANITIZER_OTHER_TAPI_FLAGS);
 OTHER_TAPI_FLAGS = $(inherited) $(WK_COMMON_OTHER_TAPI_FLAGS);
 
+// Right now, SPI analysis is off by default. Set WK_AUDIT_SPI to YES in local overrides to enable checking.
```
Nit: call it SPI auditing here too?

burg commented on Source/WTF/Scripts/audit-spi.sh:
```diff
@@ -0,0 +1,26 @@
+[ -f "${program}" ] || program="${SRCROOT}/../../Tools/Scripts/audit-spi"
+program=$(realpath "${program}")
+
+# Xcode doesn't expose the name of the discovered dependency file, but by convention, it is
```
💀

gsnedders commented on Configurations/WebKitProjectPaths.xcconfig:
```diff
@@ -98,6 +98,10 @@ WK_VERSIONED_SDK_ADDITIONS_PATH = $(WK_VERSIONED_SDK_ADDITIONS_PATH_$(USE_INTERN
 WK_VERSIONED_SDK_ADDITIONS_PATH_ = $(WK_WORKSPACE_DIR)/WebKitLibraries/SDKs/$(PLATFORM_NAME)$($(DEPLOYMENT_TARGET_SETTING_NAME):base:base).0-additions.sdk;
 WK_VERSIONED_SDK_ADDITIONS_PATH_YES = $(WK_WORKSPACE_DIR)/WebKitLibraries/SDKs/$(PLATFORM_NAME)$($(DEPLOYMENT_TARGET_SETTING_NAME):base:base).0.internal-additions.sdk;
 
+// Partial SDK DBs used for analysis are checked in to WebKitLibraries, but the
```
Nit: another place where we should call it auditing, to be consistent.

gsnedders commented on Tools/Scripts/libraries/webkitapipy/pyproject.toml:
```diff
@@ -0,0 +1,54 @@
+    'Topic :: Software Development :: Libraries :: Python Modules',
+]
+license = 'BSD-2-Clause'
+dependencies = ['PyYAML']
```
Let's be explicit about what versions we support:

```suggestion
dependencies = ['PyYAML>=5.3.1,<7']
```

gsnedders commented on Tools/Scripts/libraries/webkitapipy/webkitapipy/tbd.py:
```diff
@@ -0,0 +1,61 @@
+        yaml_tag = '!tapi-tbd'
+
+        @classmethod
+        def load_from(cls, file):
```
Why is this on this object? That feels very surprising? Especially when it's a class method that never refers `cls`.

gsnedders commented on Tools/Scripts/libraries/webkitapipy/webkitapipy/tbd.py:
```diff
@@ -0,0 +1,61 @@
+                    yield tbd
+    TBD = _PyYAMLTBD
+
+except ImportError:
```
Do we actually care about this working if PyYAML isn't installed? If we believe the `pyproject.toml` file, PyYAML is a requirement. Otherwise, we should change the dependencies, potentially adding `tbd` as an extra, which requires PyYAML?

emw-apple replied:
We care about the being able to use the underlying functionality (being able to load TBDs from an SDK) in cloud-based environments where we can't use the autoinstaller. Point taken on the pyproject.toml spec not accurately reflecting that.

Given this and below feedback, I'm tempted to switch to an alternative implementation for TBD reading. I learned that `llvm-readtapi` (available in Xcode 16+) can convert YAML-based TBDs to a JSON representation, so by switching to it we could replace the PyYAML logic entirely.

gsnedders commented on Tools/Scripts/libraries/webkitapipy/webkitapipy/tbd.py:
```diff
@@ -0,0 +1,61 @@
+try:
+    import yaml
+
+    class _PyYAMLTBD(_TBD, yaml.YAMLObject):
```
This can all be in an `else` block (c.f. https://docs.python.org/3/reference/compound_stmts.html#the-try-statement), which will run after the `try` block if no exceptions are raised, but won't mask exceptions from this.

gsnedders commented on Tools/Scripts/libraries/webkitapipy/webkitapipy/tbd.py:
```diff
@@ -0,0 +1,61 @@
+        @classmethod
+        def load_from(cls, file):
+            with open(file) as fd:
+                for tbd in yaml.unsafe_load_all(fd):
```
I'd rather we avoided using `unsafe_load_all`.

https://stackoverflow.com/questions/43058050/creating-custom-tag-in-pyyaml makes it clear it's kinda a mess, documentation wise, but we should be able to subclass `SafeLoader` (`TBDLoader`?) and call `add_constructor` on that to register our tag.

Then we just do `yaml.load_all(fd, TBDLoader)` or similar.

gsnedders commented on Tools/Scripts/libraries/webkitapipy/pyproject.toml:
```diff
@@ -0,0 +1,54 @@
+[project.scripts]
+audit-spi = "webkitapipy.program:main"
+
+[project.urls]
```
We can also add `Repository` here, albeit only pointing at all of the WebKit repo.

gsnedders commented on Tools/Scripts/libraries/webkitapipy/pyproject.toml:
```diff
@@ -0,0 +1,54 @@
+audit-spi = "webkitapipy.program:main"
+
+[project.urls]
+Homepage = "https://github.com/WebKit/WebKit/tree/main/Tools/Scripts/libraries/webkitapipy"
```
Maybe we should add a README if we're claiming this is a homepage. 🙂

gsnedders commented on Tools/Scripts/libraries/webkitapipy/webkitapipy/tbd.py:
```diff
@@ -0,0 +1,61 @@
+
+        @classmethod
+        def load_from(cls, file):
+            with open(file) as fd:
```
We should probably open this as a binary file and let PyYAML figure out the character encoding, given YAML isn't necessarily UTF-8.

gsnedders commented on Tools/Scripts/libraries/webkitapipy/webkitapipy/macho.py:
```diff
@@ -0,0 +1,166 @@
+    exports: set[str] = set()
+    selectors: set[str] = set()
+    imports: set[str] = set()
+    selrefs: set[str] = set()
```
Are these sets not held by the class, and not each instance? That seems… bad.

We probably need to override `__new__` to create new set instances for each NamedTuple; c.f. https://stackoverflow.com/a/59206181.

Also having sets (rather than frozensets) does imply the contents of the NamedTuple aren't actually constant.

emw-apple replied:
Oh, that's indeed bad! Switching to a dataclass is probably the better approach, where I can specify a [`default_factory`](https://docs.python.org/3.9/library/dataclasses.html#default-factory-functions) and we still have the autogenerated methods.

gsnedders commented on Tools/Scripts/libraries/webkitapipy/webkitapipy/macho.py:
```diff
@@ -0,0 +1,166 @@
+    selrefs: set[str] = set()
+
+    @classmethod
+    def create(cls, binary_path: Path, *, arch: str, symbols_only=False):
```
Nit: Maybe `from_binary` or something? `create` is very generic.

gsnedders commented on Tools/Scripts/libraries/webkitapipy/webkitapipy/macho.py:
```diff
@@ -0,0 +1,166 @@
+from enum import Enum
+import sys
+
+method_to_selector = re.compile(r'[-+]\[(?P<class>\S+) (?P<selector>[^\]]+)\]')
```
Nit: maybe even `objc_method_to_selector`, to make it clearer what sort of method and selector we're talking about here?

gsnedders commented on Tools/Scripts/libraries/webkitapipy/webkitapipy/macho.py:
````diff
@@ -0,0 +1,166 @@
+                # @end
+                # ```
+                m = method_to_selector.search(line)
+                if m:
````
Do we care about lines that don't match? Do we always want to silently ignore them? It feels like a change in output would potentially cause us to stop finding methods we should.

emw-apple replied:
That's definitely a risk. Maybe we can complain about unrecognized output (similar to what I do at the bottom of the loop) when there is no match, and the line doesn't contain `@interface` / `@protocol` / `@end`.

gsnedders reviewed (changes requested):
I've got to go for today, but here's a bunch of comments for your perusal first thing in the morning.

gsnedders commented on Tools/Scripts/libraries/webkitapipy/webkitapipy/macho.py:
````diff
@@ -0,0 +1,171 @@
+                # ```
+                # 0x0A8043D8  "interruption:"
+                # ```
+                address, quoted_name = line.split(maxsplit=1)
````
Given an unexpected line without any whitespace, this would throw. Maybe that's fine?

emw-apple replied:
I think it's fine, at least without evidence that this is a real risk. I'd rather it throw an exception than semi-silently discard binary information, since that could lead to our failing to detect SPI.

gsnedders commented on Tools/Scripts/libraries/webkitapipy/webkitapipy/program.py:
```diff
@@ -0,0 +1,247 @@
+
+# Some symbols, namely ones that are low-level parts of system libraries and
+# runtimes, are implicitly available.
+ALLOWED_SYMBOLS = set((
```
Nit:

```suggestion
ALLOWED_SYMBOLS = {
```

and correspondingly `}` at the end, to make this into a set literal.

emw-apple committed:
[SPI Checking] Add build-time tool for auditing SPI use
https://bugs.webkit.org/show_bug.cgi?id=290424
rdar://135979752

Reviewed by Sam Sneddon and BJ Burg.

Introduce the `webkitapipy` library and the audit-spi tool for detecting
SPI use in Mach-O binaries via static analysis. Integrate audit-spi into
the build system but leave it off by default, for now.

BACKGROUND
==========

Binary analysis tools (dyld_info) provide a record of the symbols and
Objective-C metadata used by a Mach-O binary. For each of those records,
webkitapipy checks whether the declaration is known to be API.

To do this, it uses partial SDKDBs built from projects in the SDK.
SDKDBs are a machine-readable representation of all the symbolic and
language-level declarations in a project, which are generated by TAPI.

Since the SDK is relatively unchanging compared to local build products,
we do not attempt to produce SDKDB records for the whole SDK at build
time. In a follow-up commit, WebKit will check in partial SDKDBs
generated from supported public SDKs using metadata from Apple's
internal build system. For checking internal SDKs, Apple engineers will
use separate SDKDB records available internally.

PROCEDURE
=========

At a high level, webkitapipy has a main entrypoint (program.py) which
takes a directory of SDKDBs and binaries in a build directory, and
performs SPI analysis in three stages:

1. Build an internal cache of all symbols, classes, and selectors in
   the SDK. This representation supports fast querying and is reusable
   in incremental builds. It keeps track of input files that were used to
   seed the cache along, and updates cache information when input files
   change. It is backed by a sqlite database stored in the build
   directory.

2. Load symbol, classes, and selectors from locally-built frameworks
   and dylibs whose SPI (and IPI) are allowed to be used from the binary
   being analyzed. These are also stored in the cache.

3. Run nm and various otool invocations on the binary being analyzed to
   get its symbol table, referenced selectors, and soft-linking
   metadata. Check all of these records against the cache. Missing
   records are considered to be potential SPI and are reported.

Future work will add a fourth stage, which is to parse and consult an
SPI allowlist. The eventual goal is to make use of un-allowlisted SPI a
build error, requiring committers to keep track of any SPI usage they
introduce.

INTEGRATION
===========

In the build, WebKit's frameworks have an "Audit SPI use" build phase
that calls into audit-spi. It is controlled by a WK_ANALYZE_SPI build
setting. Currently it off by default, but will eventually be
enabled for debug iOS configurations.

IMPLEMENTATION NOTES
====================

As a new, relativly self-contained Python module, I've taken the
opportunity to adopt type hints in webkitapipy. The code typechecks
successfully with mypy (`python3 -m mypy -p webkitapipy
--check-untyped-defs`). It is also fully PEP8 compliant.

A follow-up change will add a unit test which runs mypy on the module. I
hope other contributors find it helpful to adopt type hinting in their
libraries and make it easier to catch bugs ahead of time.

* Configurations/CommonBase.xcconfig:
* Configurations/SPIAnalysis.xcconfig: Added.
* Source/JavaScriptCore/JavaScriptCore.xcodeproj/project.pbxproj:
* Source/WTF/Scripts/audit-spi-if-needed.sh: Added. The Xcode frontend
  to audit-spi, which checks build settings and invokes the tool from
  the build environment. It is present in production builds (i.e.
  outside of the Tools/ directory) so that the script phases have
  something to call into.
* Source/WTF/WTF.xcodeproj/project.pbxproj:
* Source/WebCore/Configurations/WebCore.xcconfig:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebGPU/WebGPU.xcodeproj/project.pbxproj:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKitLegacy/WebKitLegacy.xcodeproj/project.pbxproj:
* Tools/Scripts/audit-spi: Added.
* Tools/Scripts/webkitpy/__init__.py: Move the PyYAML autoinstall
  dependency here, since it is no longer just used for buildbot.
* Tools/Scripts/libraries/webkitapipy/__init__.py: Added.
* Tools/Scripts/libraries/webkitapipy/program.py: Added. Implementation
  of the audit-spi tool. Handles argument parsing, loading the SDKDB
  and binary report, and reporting issues.
(TSVReporter): Very simple base class for reporting missing API as
tabular data. Will be subclasses and/or replaced in the future with
implementation that manages the SPI allowlist.
(TSVReporter.__init__):
(TSVReporter.missing_selector):
(TSVReporter.missing_class):
(TSVReporter.missing_symbol):
(TSVReporter.finished):
(main):
* Tools/Scripts/libraries/webkitapipy/macho.py: Added. Interface for
  extracting symbol and ObjC metadata from binaries and producing an
  APIReport.
(APIReport):
(APIReport.create):
(APIReport._populate_from_dyld_info):
* Tools/Scripts/libraries/webkitapipy/sdkdb.py: Added. The
  implementation of the sqlite-backed SDKDB cache.
(sdkdb):
(sdkdb.__init__):
(sdkdb._initialize_db):
(sdkdb.__del__):
(sdkdb.__enter__):
(sdkdb.__exit__):
(sdkdb._cache_hit_preparing_to_insert):
(sdkdb.add_partial_sdkdb):
(sdkdb.add_symbols):
(sdkdb.add_tbd):
(sdkdb.symbol):
(sdkdb.objc_class):
(sdkdb.objc_class.WHERE):
(sdkdb.objc_selector):
(sdkdb.stats):
(sdkdb._add_objc_interface):
(sdkdb._add_symbol):
(sdkdb._add_objc_class):
(sdkdb._add_objc_class.VALUES):
(sdkdb._add_objc_selector):
* Tools/Scripts/libraries/webkitapipy/tbd.py: Added. In addition to
  loading partial SDKDBs generated from the SDK, there are some
  libraries in the OS which are implicitly treated as API, such as
  libSystem. For these libraries, webkitapipy imports symbol
  information directly from TBDs in the SDK.
(ExportList):
(ExportList.from_json):
(TBD):
(TBD.from_json):
(TBD.from_file): Use llvm-readtapi to convert a SDKDB (typically
YAML-formatted) into a JSON representation.
* Tools/Scripts/webkitpy/autoinstalled/buildbot.py: Move PyYaml
  autoinstaller declaration from here.

Canonical link: https://commits.webkit.org/293695@main
